src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/action/LogInAction.kt

```diff
@@ -4,17 +4,38 @@ import com.atlassian.performance.tools.jiraactions.api.LOG_IN
 import com.atlassian.performance.tools.jiraactions.api.WebJira
 import com.atlassian.performance.tools.jiraactions.api.measure.ActionMeter
 import com.atlassian.performance.tools.jiraactions.api.memories.UserMemory
+import org.openqa.selenium.By
+import java.util.concurrent.ConcurrentHashMap
 
 class LogInAction(
     private val jira: WebJira,
     private val meter: ActionMeter,
     private val userMemory: UserMemory
 ) : Action {
+    private val loggedInUsers = ConcurrentHashMap.newKeySet<String>()
```

Comment:
Each VU has it's own copy of `LogInAction`, or more specifically, each calls `Scenario.getLogInAction` - they don't talk to each other by default. The `Scenario` impl can choose to return new copy each time or a shared one. By default `Scenario` creates new copies.
So right now this state is not useful, it will contain zero or one elements.

Also, instead of making VUs coordinate with one another, maybe we could run `dismissPopups` regardless? It's idempotent, right?

Reply:
So maybe I should make it static? It is idempotent, and I wouldn't have an issue if this was a functional test. But for a performance test I would prefer if we didn't do (too many) superfluous actions.

Reply:
Sounds like premature optimization. How often do we log in? How expensive is the dismiss code?

Reply:
Yeah. Maybe. I think Selenium usage requires a bit of a mindset change. Premature optimization avoidance is good for places where we have perfomance tests. Nobody tests the performance of the performance tool :-)

Reply:
I mean, I can remove it if you want. It's just that I think we should really be very careful whenever we interact with RemoteWebElements in JPT code as a rule.

Reply:
OTOH we should also avoid unnecessary locks across threads, which were generally thread-confined so far. And each piece of additional complexity risks bugs, left to be discovered in a few months.

> Nobody tests the performance of the performance tool :-)

Well we could 😛 At least from time to time.
Perhaps we could reuse `com.atlassian.performance.tools.infrastructure.api.profiler.AsyncProfiler` around `SshVirtualUsers`. It'd require `jira-actions` tests to switch from `docker-infrastructure` `infrastructure` + `ssh-ubuntu`.

From a pragmatic point of view:
* this optimization works only when VUs share logins (e.g. not for `RestUserGenerator`)
* it reduces `dismissPopups` calls by VU count minus VU node count
* the biggest tests would save ~1000 such calls, but they do use `RestUserGenerator` or similar
* in medium/small tests it would save ~20 such calls

IMHO the savings are not worth the complexity

Reply:
removed